src/memory-model.md

```diff
@@ -1,5 +1,45 @@
 # Memory model
 
-Rust does not yet have a defined memory model. Various academics and industry professionals
-are working on various proposals, but for now, this is an under-defined place
-in the language.
+r[memory]
+
+The Memory Model of Rust is incomplete and not fully decided. The following is some of the detail worked out so far.
+
+## Bytes
+
+r[memory.byte]
+
+r[memory.byte.intro]
+The most basic unit of memory in Rust is a byte. All values in Rust are computed from 0 or more bytes read from an allocation.
+
+> [!NOTE]
+> While bytes in Rust are typically lowered to hardware bytes, they may contain additional values,
+> such as being uninitialized, or storing part of a pointer.
+
+r[memory.byte.init]
+Each byte may be initialized, and contain a value of type `u8`, as well as an optional pointer fragment. When present, the pointer fragment carries [provenance][type.pointer.provenance] information.
+
+r[memory.byte.uninit]
+Each byte may be uninitialized.
+
+> [!NOTE]
+> Uninitialized bytes do not have a value and do not have a pointer fragment.
+
+## Value Encoding
+
+r[memory.encoding]
+
+r[memory.encoding.intro]
+Each type in Rust has 0 or more values, which can have operations performed on them
+
+> [!NOTE]
+> `0u8`, `1337i16`, and `Foo{bar: "baz"}` are all values
+
+r[memory.encoding.op]
+Each value of a type can be encoded into a sequence of bytes, and decoded from a sequence of bytes, which has a length equal to the size of the type.
+The operation to encode or decode a value is determined by the representation of the type.
```

Comment:
The representation relation does not determine encode/decode, since there can be more than one representation of a value -- encode makes the choice of which representation to pick. So it's the other way around, encode/decode determine the representation relation.

I would phrase this to be more centered around types, not values. Something like:

Each type defines an operation to *decode* a sequence of bytes into a value, and to *encode* a value into a sequence of bytes. When a sequence of bytes decodes to a value, we say that it *represents* that value. A value that is valid for a given type may be represented by multiple different sequences of bytes, and a sequence of bytes may represent 0 or 1 values of any given type.

Reply:
The only thing is that most of the other chapters will define just representation to get both at once. I'd like to preserve this ability.

Reply:
Although in most cases, simply defining the decoding should suffice as `decode(T, encode(T, val))` is defined to round-trip `val` later.

Reply:
> The only thing is that most of the other chapters will define just representation to get both at once. I'd like to preserve this ability.

Formally speaking, that's sloppy. If we say that every type has encode and decode, we have to say explicitly how they are defined anywhere.

> Although in most cases, simply defining the decoding should suffice as decode(T, encode(T, val)) is defined to round-trip val later.

So then I would suggest that these types have something like
```
r[<>.repr.decode]
...

r[<>.repr.encode]
The encoding of a value is the unique sequence of bytes that decodes back to the value
```

However, note that for most types there is no unique such sequence (since decode often ignores provenance). With more setup we could define the encoding as the unique minimal sequence of bytes that decodes back to the value, where "minimal" uses an ordering that sets "uninit < init(no-provenance) < init(with-provenance)".